### PR TITLE
Feature/207 visualize molecular orbitals

### PR DESCRIPTION
**Pull request by Michel-Heinz**
The molecular orbitals are almost ready. I only need to write all the tests.

## Summary by CodeRabbit

- **New Features**
	- Introduced a dialog interface for displaying molecular orbitals, enhancing user interaction and visualization capabilities.
	- Added a new action in the user interface to display molecular orbitals.
	- New files for molecular orbital data in Molden format for hydrogen and water molecules.

- **Bug Fixes**
	- Improved handling of input files in the importers, enhancing robustness and flexibility.

- **Documentation**
	- Added docstrings to clarify module purposes and functionality.

- **Refactor**
	- Renamed and reorganised various classes and methods for better clarity and consistency.
	- Updated the handling of basis functions and molecular orbital calculations.

- **Tests**
	- Enhanced test cases to align with recent changes in functionality and improve precision in assertions.

### CODE REVIEW

**Comment by Michel-Heinz** [comment]
@adrianusler Do you know why these tests are failing only for different OSs? These are the tests, that test the cylinder and sphere drawing routines :)

**Comment by ab5424** [comment]
> @adrianusler Do you know why these tests are failing only for different OSs? These are the tests, that test the cylinder and sphere drawing routines :)

@Michel-Heinz The gui and rendering tests are skipped on windows.

**Comment by Michel-Heinz** [comment]
On my local PC the tests run with no problem...

**Comment by ab5424** [comment]
@Michel-Heinz Which `numpy` version do you have installed on your PC? It it is <2 then I know why the testa are failing.

**Comment by Michel-Heinz** [comment]
I got it, I can reproduce the failure now

**Comment by Michel-Heinz** [comment]
Great only codecov missing. I got my work cut out.

**Comment by ab5424** [comment]
@Michel-Heinz I've reverted commit [e63e585](https://github.com/Molara-Lab/Molara/commit/e63e58571a0e1b500ff76ef2372e586d17ee1a40) (except for non-formatting changes). That commit was polluting the change history because `black` introduced magic trailing commas everywhere, which lead to a lot of changes throughout the code base. And since `ruff` thinks they were introduced on purpose it won't remove them. If you want to format the code it s better to run `ruff format` (otherwise the increase of line breaks in combination with untested functions means you'll have a hard time keeping the coverage on it's previous level).

**Comment by ab5424** [comment]
@coderabbitai review

**Comment by Michel-Heinz** [comment]
> @Michel-Heinz I've reverted commit [e63e585](https://github.com/Molara-Lab/Molara/commit/e63e58571a0e1b500ff76ef2372e586d17ee1a40) (except for non-formatting changes). That commit was polluting the change history because `black` introduced magic trailing commas everywhere, which lead to a lot of changes throughout the code base. And since `ruff` thinks they were introduced on purpose it won't remove them. If you want to format the code it s better to run `ruff format` (otherwise the increase of line breaks in combination with untested functions means you'll have a hard time keeping the coverage on it's previous level).

So no more `pre-commit run --all-files` ? or just never do black alone?

**Comment by ab5424** [comment]
> So no more `pre-commit run --all-files` ? or just never do black alone?

`pre-commit run --all-files`is perfectly fine, as is `ruff format .` The only problem is `black` because it can't read the settings for the ruff formatter we have in our `pyproject.toml`

**Comment by Michel-Heinz** [comment]
Do not merge yet! There are still tests missing, codecov just does not know...

**Comment by Michel-Heinz** [comment]
@ab5424 I did not run black, but somehow I changed over 14000 lines now?

**Comment by ab5424** [comment]
> @ab5424 I did not run black, but somehow I changed over 14000 lines now?

Well, your molden files have a few thousand lines, so...

File: src/molara/eval/populationanalysis.py
Comment (Michel-Heinz) [reply src/molara/eval/populationanalysis.py:73]: there are only cartesian ones or spherical ones.

File: src/molara/eval/aos.pyx
Comment (Michel-Heinz) [reply src/molara/eval/aos.pyx:67]: Nice one 👍

File: src/molara/eval/generate_voxel_grid.pyx
Comment (Michel-Heinz) [reply src/molara/eval/generate_voxel_grid.pyx:66]: not time critical

File: src/molara/eval/marchingcubes.py
Comment (Michel-Heinz) [reply src/molara/eval/marchingcubes.py:138]: This will be rewritten in cython at some stage

File: src/molara/eval/marchingcubes.py
Comment (Michel-Heinz) [reply src/molara/eval/marchingcubes.py:173]: Nice one

File: src/molara/eval/marchingcubes.py
Comment (Michel-Heinz) [reply src/molara/eval/marchingcubes.py:172]: nice

**Comment by Michel-Heinz** [comment]
2 of coderabbits suggestions fixed a bug, that I had some issues with, very nice 👍

**Comment by ab5424** [comment]
I opened the file `o2.molden`, went on "Display MOs", clicked on "Display Orbital" but got the following error:

```
Traceback (most recent call last):
  File "D:\GitHub\Molara\src\molara\gui\mos_dialog.py", line 344, in visualize_orbital
    voxel_grid = generate_voxel_grid(
  File "src\\molara\\eval\\generate_voxel_grid.pyx", line 11, in molara.eval.generate_voxel_grid.generate_voxel_grid
    cpdef generate_voxel_grid(
  File "src\\molara\\eval\\generate_voxel_grid.pyx", line 33, in molara.eval.generate_voxel_grid.generate_voxel_grid
    cdef long[:,:] orbital_ijks = npc.ndarray(shape=(number_of_aos, 3), dtype=np.int64)
ValueError: Buffer dtype mismatch, expected 'long' but got 'long long'
```

**Comment by ab5424** [comment]
And when I click on "Display Box", I get the following:

```
Traceback (most recent call last):
  File "D:\GitHub\Molara\src\molara\gui\structure_widget.py", line 219, in paintGL
    self.renderer.draw_scene(self.camera, self.bonds)
  File "D:\GitHub\Molara\src\molara\rendering\rendering.py", line 497, in draw_scene
    glBindVertexArray(sphere["vao"])
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\error.py", line 230, in glCheckError
    raise self._errorClass(
OpenGL.error.GLError: GLError(
        err = 1282,
        description = b'Der Vorgang ist ung\xfcltig.',
        baseOperation = glBindVertexArray,
        cArguments = (np.uint32(3),)
)
```

**Comment by Michel-Heinz** [comment]
> I opened the file `o2.molden`, went on "Display MOs", clicked on "Display Orbital" but got the following error:
> 
> ```
> Traceback (most recent call last):
>   File "D:\GitHub\Molara\src\molara\gui\mos_dialog.py", line 344, in visualize_orbital
>     voxel_grid = generate_voxel_grid(
>   File "src\\molara\\eval\\generate_voxel_grid.pyx", line 11, in molara.eval.generate_voxel_grid.generate_voxel_grid
>     cpdef generate_voxel_grid(
>   File "src\\molara\\eval\\generate_voxel_grid.pyx", line 33, in molara.eval.generate_voxel_grid.generate_voxel_grid
>     cdef long[:,:] orbital_ijks = npc.ndarray(shape=(number_of_aos, 3), dtype=np.int64)
> ValueError: Buffer dtype mismatch, expected 'long' but got 'long long'
> ```
> 
> Edit: Windows 11, python3.10, numpy 2.1.0

This should be fixed now.

**Comment by ab5424** [comment]
> This should be fixed now.

@Michel-Heinz No it's unfortunately still there (I did reinstalled molara).

**Comment by ab5424** [comment]
@Michel-Heinz Now I get the following:

```
dict_keys([<class 'bytes'>, <class 'OpenGL.converters.c_long_Array_1'>, <class 'numpy.ndarray'>, <class 'ctypes.c_char_Array_84'>, <class 'pyrr.objects.vector3.Vector3'>, <class 'ctypes.c_void_p'>, <class 'int'>, <class 'ctypes.c_ulong'>, <class 'NoneType'>])
dict_keys([<class 'bytes'>, <class 'OpenGL.converters.c_long_Array_1'>, <class 'numpy.ndarray'>, <class 'ctypes.c_char_Array_84'>, <class 'pyrr.objects.vector3.Vector3'>, <class 'ctypes.c_void_p'>, <class 'int'>, <class 'ctypes.c_ulong'>, <class 'NoneType'>])
Traceback (most recent call last):
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\latebind.py", line 43, in __call__
    return self._finalCall( *args, **named )
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\wrapper.py", line 689, in wrapperCall
    pyArgs = tuple( calculate_pyArgs( args ))
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\wrapper.py", line 450, in calculate_pyArgs
    yield converter(args[index], self, args)
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\converters.py", line 135, in __call__
    return self.function( incoming )
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\arrays\arraydatatype.py", line 171, in asArray
    return cls.getHandler(value).asArray(value, typeCode or cls.typeConstant)
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\arrays\arraydatatype.py", line 59, in __call__
    raise TypeError(
TypeError: ('No array-type handler for type numpy.uintc (value: np.uint32(10)) registered', <OpenGL.converters.CallFuncPyConverter object at 0x000001949D4AC0A0>)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\GitHub\Molara\src\molara\gui\mos_dialog.py", line 229, in closeEvent
    self.remove_orbitals()
  File "D:\GitHub\Molara\src\molara\gui\mos_dialog.py", line 381, in remove_orbitals
    self.parent().structure_widget.renderer.remove_polygon(orb)
  File "D:\GitHub\Molara\src\molara\rendering\rendering.py", line 130, in remove_polygon
    glDeleteBuffers(1, buffer)
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\latebind.py", line 47, in __call__
    return self._finalCall( *args, **named )
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\wrapper.py", line 689, in wrapperCall
    pyArgs = tuple( calculate_pyArgs( args ))
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\wrapper.py", line 450, in calculate_pyArgs
    yield converter(args[index], self, args)
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\converters.py", line 135, in __call__
    return self.function( incoming )
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\arrays\arraydatatype.py", line 171, in asArray
    return cls.getHandler(value).asArray(value, typeCode or cls.typeConstant)
  File "D:\PycharmProjects\Molara-venv\lib\site-packages\OpenGL\arrays\arraydatatype.py", line 59, in __call__
    raise TypeError(
TypeError: ('No array-type handler for type numpy.uintc (value: np.uint32(10)) registered', <OpenGL.converters.CallFuncPyConverter object at 0x000001949D4AC0A0>)
```

Btw, I'd be fine with merging this for now and fixing the windows issues in a separate PR if that'd work for you.

**Comment by Michel-Heinz** [comment]
@ab5424 , @adrianusler , @GereonFeldmann  I support Alexs suggestion of merging and then testing windows. I will write here as soon as I wrote all the tests. Than we can merge this!

**Comment by Michel-Heinz** [comment]
I am happy now, with the tests and the whole project. From my side this is ready to merge. As soon as it is merged an issue for the windows issues should be posted. 💯

**Comment by adrianusler** [comment]
@Michel-Heinz I can't run Molara in the `feature/207` branch. I get this error:
```bash
Traceback (most recent call last):
  File ".../venv/bin/molara", line 7, in <module>
    from molara.__main__ import main
  File ".../src/molara/__main__.py", line 16, in <module>
    from molara.gui.main_window import MainWindow
  File ".../src/molara/gui/main_window.py", line 12, in <module>
    from molara.gui.builder import BuilderDialog
  File ".../src/molara/gui/builder.py", line 14, in <module>
    from molara.structure.molecule import Molecule
  File ".../src/molara/structure/molecule.py", line 9, in <module>
    from molara.structure.molecularorbitals import MolecularOrbitals
  File ".../src/molara/structure/molecularorbitals.py", line 10, in <module>
    from molara.eval.mos import calculate_mo_cartesian
ModuleNotFoundError: No module named 'molara.eval.mos'
```

**Comment by ab5424** [comment]
> @Michel-Heinz I can't run Molara in the `feature/207` branch. I get this error:
> 
> ```shell
> Traceback (most recent call last):
>   File ".../venv/bin/molara", line 7, in <module>
>     from molara.__main__ import main
>   File ".../src/molara/__main__.py", line 16, in <module>
>     from molara.gui.main_window import MainWindow
>   File ".../src/molara/gui/main_window.py", line 12, in <module>
>     from molara.gui.builder import BuilderDialog
>   File ".../src/molara/gui/builder.py", line 14, in <module>
>     from molara.structure.molecule import Molecule
>   File ".../src/molara/structure/molecule.py", line 9, in <module>
>     from molara.structure.molecularorbitals import MolecularOrbitals
>   File ".../src/molara/structure/molecularorbitals.py", line 10, in <module>
>     from molara.eval.mos import calculate_mo_cartesian
> ModuleNotFoundError: No module named 'molara.eval.mos'
> ```

@adrianusler Please check if the folder eval is still capitalised on your file system, i.e. `src/molara/Eval`. Git sometimes has problems with that.

**Comment by adrianusler** [comment]
> @adrianusler Please check if the folder eval is still capitalised on your file system, i.e. `src/molara/Eval`. Git sometimes has problems with that.

I also suspected something of the sort, but the problem is still there in a fresh clone. So it can't really be an artefact of my local file structure, right?

**Comment by adrianusler** [comment]
> > @adrianusler Please check if the folder eval is still capitalised on your file system, i.e. `src/molara/Eval`. Git sometimes has problems with that.
> 
> I also suspected something of the sort, but the problem is still there in a fresh clone. So it can't really be an artefact of my local file structure, right?
> 

My bad, I should have re-run pip install on the new branch...

**Comment by adrianusler** [comment]
I tried opening the water molden file and displaying the molecular orbitals. When I want to display the box or the orbitals, I get this error message:
```bash
Error calling Python override of QOpenGLWidget::paintGL(): Traceback (most recent call last):
  File ".../src/molara/gui/structure_widget.py", line 219, in paintGL
    self.renderer.draw_scene(self.camera, self.bonds)
  File ".../src/molara/rendering/rendering.py", line 512, in draw_scene
    glBindVertexArray(polygon["vao"])
  File ".../venv/lib/python3.10/site-packages/OpenGL/error.py", line 230, in glCheckError
    raise self._errorClass(
OpenGL.error.GLError: GLError(
	err = 1282,
	description = b'invalid operation',
	baseOperation = glBindVertexArray,
	cArguments = (np.uint32(8),)
)
```

**Comment by Michel-Heinz** [comment]
Ready to merge!

**Review by adrianusler — APPROVED** [review]
What a great new feature. Thank you, @Michel-Heinz!

**Pull request merged**
